charm-slurmd/metadata.yaml

```diff
@@ -33,3 +33,16 @@ requires:
 peers:
   slurmd-peer:
     interface: slurmd-peer
+
+resources:
+  singularity-deb:
+    type: file
+    filename: singularity-ce_3.10.2-focal_amd64.deb
```

Comment:
shouldn't the charm pull these .deb files? Sorry if it's the case but as far as I understood, the charm expects it to exist previously?

Reply:
Currently I need to to download the .deb (or .rpm) before deploying the charm. And then pass the file via `--resource singularity-deb=path/to/singularity-ce_3.10.2-focal_amd64.deb`. Would you have any suggestion of a better way to do this?If so, that would be great.

Reply:
I'd say maybe passing the url for the official .deb and download it in the runtime. What do you think? I thought something like:

```python
import urllib.request
urllib.request.urlretrieve("https://github.com/sylabs/singularity/releases/download/v3.10.2/singularity-ce-3.10.2-1.el7.x86_64.rpm", "singularity-ce-3.10.2-1.rpm")
```

Then the url would need to be passed of course

Reply:
I like the idea. I think it's better to have something as automatic as possible. I just wonder where (or when) the download should be done. The ideia of having the .deb and .rpm files as resources was to avoid multiple downloads in runtime. I am still figuring out how the charms deployment process works. So it would be nice if you could tell me the best place to download the file in order to optimize the process. Thanks in advance.

Reply:
I think we can download from github itself. The link I sent previously it's from the official github release

Reply:
https://github.com/sylabs/singularity/releases/tag/v3.10.2

Reply:
I am sorry, I was not clear. I mean, the place in the code where we should start downloading the file. 

Currently I download it from https://github.com/sylabs/singularity/releases/download/v3.10.2/singularity-ce_3.10.2-focal_amd64.deb already (but before deployment). Also, the installation process is going to be done in the slurm-ops-manager (as @heitorPB 's suggestion  )

Reply:
hmm I see. I think you can have a `install` method in the class in define in the slurm-ops-manager. It's not clear to me as well right now. Perhaps we're going to have a better picture of that when you start to sketch the change Heitor suggested

Reply:
Sure, let's discuss this after I finish the changes.

Reply:
> shouldn't the charm pull these .deb files?

The best practice here is to supply the binaries via juju resources. This adds the benefits of off-line deploys as well as a better control of the installation procedure.

Reply:
+1

Reply:
> > shouldn't the charm pull these .deb files?
> 
> The best practice here is to supply the binaries via juju resources. This adds the benefits of off-line deploys as well as a better control of the installation procedure.

off-line deploys won't work with these charms because it is needed to download the slurm rpm packages from Omnivector's rpm repo. About the control: adding the URl would allow the same control.

The benefit I see in passing the URL is no extra step is needed in the user end. The charm can have a default URL depending on the system and the user could also pass the URL, so, it's a very customized way.

Reply:
I can't see how having to install the binary first is beneficial

Reply:
> off-line deploys won't work with these charms because it is needed to download the slurm rpm packages from Omnivector's rpm repo.

That is not true. We can have off-line deploys _and_ software from a Software Repository. In fact, we do that :) There are many systems that caches the packages from a public repo in a local server, in an automatic and transparent way.


> About the control: adding the URl would allow the same control.

Using an URL would not give us the same control. This strategy reduces the scalability of our deploys, as each unit would need to download something from an external server. And there's no guarantee that that request will complete. An external network outage could prevent us from doing a local deploy. These are some of the scenarios that we need to anticipate when working with infrastructure and provide a solution in advance.

By providing Singularity as a Juju resource, we can control the lifecycle operations with more detail and granularity.


Juju resource is the recommended way in this case. I am closing this discussion as it is getting off-topic and we should focus instead on reviewing this PR. If you want to understand more about Juju resources I encourage you to open a topic in the [Juju forums](https://discourse.charmhub.io/) and we can discuss it there.

Reply:
well, you showed good points. Indeed, for caching you need to download once so the off-line deploy is not 100% off-line. And again, external network outage would also not allow downloading the rpm locally... I still cannot see how can it be beneficial having to have a previous step. This is bad in **any** installation process. I remember specially the times of Tensorflow 1.x in which it was needed to have some installations outside the software depending on what you would be doing, it was terrible. But as you said, let's move on.